fs/fs_zenfs.cc

```diff
@@ -1016,8 +1155,258 @@ Status ZenFS::Mount(bool readonly) {
+      if (s.IsNotFound()) {
+        Warn(logger_,
+             "Did not find a valid snapshot, trying next snapshot zone. "
+             "Error: %s", s.ToString().c_str());
+        continue;
+      }
+
+      Error(logger_, "Snapshot corruption. Error: %s", s.ToString().c_str());
+      return s;
+    }
+
+    snapshot_recovered_index = i;
+    snapshot_recovery_ok = true;
+    snapshot_log_ = std::move(log);
+    break;
+  }
+
+  if (!snapshot_recovery_ok) {
+    return Status::IOError("Failed to mount filesystem due to "
+                           "snapshot recovery");
+  }
+
+  Info(logger_, "Recovered from snapshot zone: %d",
+         (int)valid_zones_snapshot[snapshot_recovered_index]->GetZoneNr());
+  snapshot_superblock_ =
+    std::move(valid_superblocks_snapshot[snapshot_recovered_index]);
+
+  /* Free up old snapshot zones, to get ready to roll */
+  for (const auto& sm : seq_map_snapshot) {
+    uint32_t i = sm.second;
+    /* Don't reset the current snapshot zone */
+    if (i != snapshot_recovered_index) {
+      /* snapshot zones are not marked as having valid data, so they can be
+       * reset */
+      valid_logs_snapshot[i].reset();
+    }
+  }
+
+  /* Find all valid superblocks in metazones*/
+  for (const auto z : metazones) {
+    std::unique_ptr<ZenMetaLog> log;
+    std::string scratch;
+    Slice super_record;
+
+    log.reset(new ZenMetaLog(zbd_, z));
+
+    if (!log->ReadRecord(&super_record, &scratch).ok()) continue;
+
+    if (super_record.size() == 0) continue;
+
+    std::unique_ptr<Superblock> super_block;
+
+    super_block.reset(new Superblock());
+    s = super_block->DecodeFrom(&super_record);
+    if (s.ok()) s = super_block->CompatibleWith(zbd_);
+    if (!s.ok()) return s;
+
+    Info(logger_, "Found OK superblock in zone %lu seq: %u\n", z->GetZoneNr(),
+         super_block->GetSeq());
+
+    seq_map_metadata.push_back(std::make_pair(super_block->GetSeq(),
+                                 seq_map_metadata.size()));
+    valid_superblocks_metadata.push_back(std::move(super_block));
+    valid_logs_metadata.push_back(std::move(log));
+    valid_zones_metadata.push_back(z);
+  }
+
+  if (!seq_map_metadata.size()) {
+    return Status::NotFound("No valid superblock found from metadata zones");
+  }
+
+  /* Sort superblocks by descending sequence number in metadata zones*/
+  std::sort(seq_map_metadata.begin(), seq_map_metadata.end(),
+            std::greater<std::pair<uint32_t, uint32_t>>());
+
+  bool metadata_recovery_ok = false;
+  unsigned int metadata_recovered_index = 0;
+
+  /* Recover from the zone with the highest superblock sequence number.
+     If that fails go to the previous as we might have crashed when rolling
+     metadata zone.
+  */
+  for (const auto& sm : seq_map_metadata) {
+    uint32_t i = sm.second;
+    std::string scratch;
+    std::unique_ptr<ZenMetaLog> log = std::move(valid_logs_metadata[i]);
+
+    s = RecoverFromMetaZone(log.get());
+    if (!s.ok()) {
```

Comment:
Question: in MKFSV2, we don't write an empty metadata record in the metadata zone. However, in `RecoverFromMetaZone`, we only have two valid states `kFileUpdate` and `kFileDeletion` to find a non-empty metadata. So from empty or formatted metazones, this condition `s.ok()` is always false. Thus we can't initiate `meta_log_`, `metadata_recovered_index`, `metadata_recovery_ok`, or reach lines 1356 to 1359. 

Shall we add an empty metadata record when formatting metadata zone? This will add an another state, for example `kEmptyMetadata` to `ZenFSTag` and add some other helper functions like `PersistEmptyMetadata`